src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1413 @@
+        {
+        }
+
+        public TypeInferenceContext(PowerShell powerShell)
+        {
+            _powerShell = powerShell;
+            Helper = new PowerShellExecutionHelper(powerShell);
+
+        }
+
+        public TypeDefinitionAst CurrentTypeDefinitionAst { get; set; }
+        public TypeInferenceRuntimePermissions RuntimePermissions { get; set; }
+        internal ExecutionContext ExecutionContext => _powerShell.Runspace.ExecutionContext;
+
+        internal IEnumerable<PSTypeName> InferType(Ast ast, TypeInferenceVisitor visitor)
+        {
+            var res = ast.Accept(visitor);
+            if (res == null)
+            {
+                return EmptyPSTypeNameArray;
+            }
+            return (IEnumerable<PSTypeName>)res;
+        }
+
+        public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst expression, out PSTypeName typeName)
+        {
+            typeName = null;
+            if (!RuntimePermissions.HasFlag(TypeInferenceRuntimePermissions.AllowSafeEval)) return false;
+            object value;
+            return expression != null &&
+                   SafeExprEvaluator.TrySafeEval(expression, ExecutionContext, out value) &&
+                   TryGetRepresentativeTypeNameFromValue(value, out typeName);
+        }
+
+        internal static bool TryGetRepresentativeTypeNameFromValue(object value, out PSTypeName type)
+        {
+            type = null;
+            if (value != null)
+            {
+                var list = value as IList;
+                if (list != null && list.Count > 0)
+                {
+                    value = list[0];
+                }
+                value = PSObject.Base(value);
+                if (value != null)
+                {
+                    type = new PSTypeName(value.GetType());
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public IList<object> GetMembersByInferredType(PSTypeName typename, bool @static, Func<object, bool> filter)
+        {
+            List<object> results = new List<object>();
+
+            Func<object, bool> filterToCall = filter;
+            if (typename.Type != null)
+            {
+                if (CurrentTypeDefinitionAst == null || CurrentTypeDefinitionAst.Type != typename.Type)
+                {
+                    if (filterToCall == null)
+                        filterToCall = o => !IsMemberHidden(o);
+                    else
+                        filterToCall = o => !IsMemberHidden(o) && filter(o);
+                }
+                IEnumerable<Type> elementTypes;
+                if (typename.Type.IsArray)
+                {
+                    elementTypes = new[] { typename.Type.GetElementType() };
+                }
+                else
+                {
+                    elementTypes = typename.Type.GetInterfaces().Where(
```

Comment:
Code in this file may be executed on every keystroke in some scenarios, so it's better avoid LINQ and extensive lambda creation.

https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#performance-considerations